lib/host/wasi/inode-win.cpp

```diff
@@ -145,20 +851,196 @@ WasiExpect<void> INode::fdRead(Span<Span<uint8_t>> IOVs,
   return {};
 }
 
-WasiExpect<void> INode::fdReaddir(Span<uint8_t>, __wasi_dircookie_t,
-                                  __wasi_size_t &) noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+WasiExpect<void> INode::fdReaddir(Span<uint8_t> Buffer,
+                                  __wasi_dircookie_t Cookie,
+                                  __wasi_size_t &Size) noexcept {
+
+  WIN32_FIND_DATAW FindData;
+  uint64_t Seek = 0;
+  wchar_t HandleFullPathW[MAX_PATH];
+  wchar_t FullPathW[MAX_PATH];
+
+  std::vector<uint8_t, boost::alignment::aligned_allocator<
+                           uint8_t, alignof(__wasi_dirent_t)>>
+      LocalBuffer;
+
+  // First get the path of the handle
+  if (unlikely(GetFinalPathNameByHandleW(Handle, HandleFullPathW, MAX_PATH,
+                                         FILE_NAME_NORMALIZED)) == FALSE) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  // Check if the path is a directory or not
+  if (!PathIsDirectoryW(HandleFullPathW)) {
+    return WasiUnexpect(__WASI_ERRNO_NOTDIR);
+  }
+
+  // WildCard to match every file/directory present in the directory
+  const wchar_t WildCard[]{L"\\*"};
+  HRESULT CombineResult =
+      PathCchCombine(FullPathW, MAX_PATH, HandleFullPathW, WildCard);
+
+  switch (CombineResult) {
+  case S_OK:
+    break;
+  case E_INVALIDARG:
+    return WasiUnexpect(__WASI_ERRNO_INVAL);
+  case E_OUTOFMEMORY:
+    return WasiUnexpect(__WASI_ERRNO_OVERFLOW);
+  default:
+    return WasiUnexpect(__WASI_ERRNO_NAMETOOLONG);
```

Comment:
Just want to know why the default case of CombineResult is `Name too long`?

Reply:
No there are four possible return values. `PATHCCH_E_FILENAME_TOO_LONG` is the case the default is representing. But this constant is not defined in `Pathcch.h` and I could not find where it was defined. So I made it the default case.
Reference: https://docs.microsoft.com/en-us/windows/win32/api/pathcch/nf-pathcch-pathcchcombine#return-value